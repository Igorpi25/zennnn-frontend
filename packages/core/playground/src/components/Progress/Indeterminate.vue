<template>
  <div class="mb-6">
    <Progress
      indeterminate
      class="mr-6"
    />
    <Progress
      indeterminate
      class="text-red-500 mr-6"
    />
    <Progress
      indeterminate
      class="text-yellow-500 mr-6"
    />
    <Progress
      indeterminate
      class="text-green-500"
    />
  </div>
  <div class="text-blue-500">
    <Progress
      :indeterminate="indeterminate"
      class="mr-6"
    />
    <Btn x-small outlined retain-focus-on-click @click="indeterminate = !indeterminate">
      Toggle
    </Btn>
  </div>
</template>

<script setup>
import { ref } from 'vue'

const indeterminate = ref(false)
</script>
