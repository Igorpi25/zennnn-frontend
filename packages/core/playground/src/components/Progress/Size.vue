<template>
  <div class="text-blue-500 -mb-6">
    <Progress
      :value="25"
      class="mr-6 mb-6"
    />
    <Progress
      :value="50"
      :width="3"
      class="mr-6 mb-6"
    />
    <Progress
      :value="80"
      :size="50"
      :width="5"
      class="mr-6 mb-6"
    />
    <Progress
      :value="70"
      :size="70"
      :width="7"
      class="mb-6"
    />
  </div>
</template>
