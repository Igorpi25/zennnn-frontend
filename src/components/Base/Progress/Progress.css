@layer components {
  .progress {
    position: relative;
    display: inline-flex;
    vertical-align: middle;
    justify-content: center;
    align-items: center;
  }
  .progress svg {
    width: 100%;
    height: 100%;
    margin: auto;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 0;
  }
  .progress--indeterminate svg {
    animation: progress-rotate 1.4s linear infinite;
    transform-origin: center center;
    transition: all .2s ease-in-out;
  }

  .progress--indeterminate .progress__overlay {
    animation: progress-dash 1.4s ease-in-out infinite;
    stroke-linecap: round;
    stroke-dasharray: 80, 200;
    stroke-dashoffset: 0px;
  }

  .progress__underlay {
    stroke: rgba(0, 0, 0, 0.1);
    z-index: 1;
  }
  .dark .progress__underlay {
    stroke: rgba(103, 103, 103, 0.4);
  }

  .progress__overlay {
    stroke: currentColor;
    z-index: 2;
    transition: all .6s ease-in-out;
  }
  .progress__info {
    font-size: 14px;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  @keyframes progress-dash {
    0% {
      stroke-dasharray: 1, 200;
      stroke-dashoffset: 0px;
    }
  
    50% {
      stroke-dasharray: 100, 200;
      stroke-dashoffset: -15px;
    }
  
    100% {
      stroke-dasharray: 100, 200;
      stroke-dashoffset: -125px;
    }
  }

  @keyframes progress-rotate {
    100% {
      transform: rotate(360deg);
    }
  }
}
