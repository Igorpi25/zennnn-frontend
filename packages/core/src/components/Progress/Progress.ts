import { h, reactive, computed, defineComponent } from 'vue'
import { convertToUnit } from 'vue-supp'

import './Progress.css'

export default defineComponent({
  name: 'Progress',

  props: {
    indeterminate: Boolean,
    rotate: {
      type: [Number, String],
      default: 0,
    },
    size: {
      type: [Number, String],
      default: 32,
    },
    width: {
      type: [Number, String],
      default: 4,
    },
    value: {
      type: [Number, String],
      default: 0,
    },
  },

  setup (props, { slots }) {
    const state: any = reactive({
      radius: 20,
      circumference: computed(() => {
        return 2 * Math.PI * state.radius
      }),
      normalizedValue: computed(() => {
        if (props.value < 0) {
          return 0
        }

        if (props.value > 100) {
          return 100
        }

        return parseFloat(props.value)
      }),
      strokeDashArray: computed(() => {
        return Math.round(state.circumference * 1000) / 1000
      }),
      strokeDashOffset: computed(() => {
        return ((100 - state.normalizedValue) / 100) * state.circumference + 'px'
      }),
      strokeWidth: computed(() => {
        return Number(props.width) / +props.size * state.viewBoxSize * 2
      }),
      svgStyles: computed(() => {
        return {
          transform: `rotate(${Number(props.rotate)}deg)`,
        }
      }),
      viewBoxSize: computed(() => {
        return state.radius / (1 - Number(props.width) / +props.size)
      }),
      styles: computed(() => {
        const s = Number(props.size)
        return {
          height: convertToUnit(s),
          width: convertToUnit(s),
        }
      }),
      classes: computed(() => {
        return {
          progress: true,
          'progress--indeterminate': props.indeterminate,
        }
      }),
    })

    const genCircle = (name: string, offset: number) => {
      return h('circle', {
        class: `progress__${name}`,
        fill: 'transparent',
        cx: 2 * state.viewBoxSize,
        cy: 2 * state.viewBoxSize,
        r: state.radius,
        'stroke-width': state.strokeWidth,
        'stroke-dasharray': state.strokeDashArray,
        'stroke-dashoffset': offset,
      })
    }

    const genSvg = () => {
      const children = [
        props.indeterminate || genCircle('underlay', 0),
        genCircle('overlay', state.strokeDashOffset),
      ]

      return h('svg', {
        style: state.svgStyles,
        xmlns: 'http://www.w3.org/2000/svg',
        viewBox: `${state.viewBoxSize} ${state.viewBoxSize} ${2 * state.viewBoxSize} ${2 * state.viewBoxSize}`,
      }, children)
    }

    const genInfo = () => {
      return h('div', {
        class: 'progress__info',
      }, slots.default?.())
    }

    return () => {
      return h('div', {
        role: 'progressbar',
        'aria-valuemin': 0,
        'aria-valuemax': 100,
        'aria-valuenow': props.indeterminate ? undefined : state.normalizedValue,
        class: state.classes,
        style: state.styles,
      }, [
        genSvg(),
        genInfo(),
      ])
    }
  },
})
