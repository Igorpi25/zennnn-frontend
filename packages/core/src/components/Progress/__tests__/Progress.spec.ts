import { mount, VueWrapper } from '@vue/test-utils'
import Progress from '../Progress'

describe('Progress.ts', () => {
  let mountFunction: (options?: Record<string, unknown>) => VueWrapper<any>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Progress, {
        ...options,
      })
    }
  })

  // Render

  it('should render component and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        value: 33,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ value: -1 })
    const htmlMinus1 = wrapper.html()

    await wrapper.setProps({ value: 0 })
    const html0 = wrapper.html()

    await wrapper.setProps({ value: 100 })
    const html100 = wrapper.html()

    await wrapper.setProps({ value: 101 })
    const html101 = wrapper.html()

    expect(htmlMinus1).toBe(html0)
    expect(html100).toBe(html101)
    expect(html0).not.toBe(html100)

    await wrapper.setProps({ value: '-1' })
    const htmlMinus1String = wrapper.html()

    await wrapper.setProps({ value: '0' })
    const html0String = wrapper.html()

    await wrapper.setProps({ value: '100' })
    const html100String = wrapper.html()

    await wrapper.setProps({ value: '101' })
    const html101String = wrapper.html()

    expect(htmlMinus1String).toBe(html0String)
    expect(html100String).toBe(html101String)
    expect(html0String).not.toBe(html100String)
  })

  it('should render component with rotate prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: { value: 33, rotate: 44 },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with size prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: { value: 33, size: 16 },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with width prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: { value: 33, width: 1 },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with indeterminate prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: { indeterminate: true },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with default slot and match snapshot', () => {
    const wrapper = mountFunction({
      props: { value: 33 },
      slots: {
        default: () => '33',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
