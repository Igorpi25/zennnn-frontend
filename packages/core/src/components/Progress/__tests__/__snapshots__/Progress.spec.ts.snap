// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress.ts should render component and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="33" class="progress" style="height: 32px; width: 32px;"><svg style="transform: rotate(0deg);" xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715">
    <circle class="progress__underlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="0"></circle>
    <circle class="progress__overlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="84.19468311620646px"></circle>
  </svg>
  <div class="progress__info"></div>
</div>
`;

exports[`Progress.ts should render component with default slot and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="33" class="progress" style="height: 32px; width: 32px;"><svg style="transform: rotate(0deg);" xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715">
    <circle class="progress__underlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="0"></circle>
    <circle class="progress__overlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="84.19468311620646px"></circle>
  </svg>
  <div class="progress__info">33</div>
</div>
`;

exports[`Progress.ts should render component with indeterminate prop and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="progress progress--indeterminate" style="height: 32px; width: 32px;"><svg style="transform: rotate(0deg);" xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715">
    <!---->
    <circle class="progress__overlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px"></circle>
  </svg>
  <div class="progress__info"></div>
</div>
`;

exports[`Progress.ts should render component with rotate prop and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="33" class="progress" style="height: 32px; width: 32px;"><svg style="transform: rotate(44deg);" xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715">
    <circle class="progress__underlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="0"></circle>
    <circle class="progress__overlay" fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="84.19468311620646px"></circle>
  </svg>
  <div class="progress__info"></div>
</div>
`;

exports[`Progress.ts should render component with size prop and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="33" class="progress" style="height: 16px; width: 16px;"><svg style="transform: rotate(0deg);" xmlns="http://www.w3.org/2000/svg" viewBox="26.666666666666668 26.666666666666668 53.333333333333336 53.333333333333336">
    <circle class="progress__underlay" fill="transparent" cx="53.333333333333336" cy="53.333333333333336" r="20" stroke-width="13.333333333333334" stroke-dasharray="125.664" stroke-dashoffset="0"></circle>
    <circle class="progress__overlay" fill="transparent" cx="53.333333333333336" cy="53.333333333333336" r="20" stroke-width="13.333333333333334" stroke-dasharray="125.664" stroke-dashoffset="84.19468311620646px"></circle>
  </svg>
  <div class="progress__info"></div>
</div>
`;

exports[`Progress.ts should render component with width prop and match snapshot 1`] = `
<div role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="33" class="progress" style="height: 32px; width: 32px;"><svg style="transform: rotate(0deg);" xmlns="http://www.w3.org/2000/svg" viewBox="20.64516129032258 20.64516129032258 41.29032258064516 41.29032258064516">
    <circle class="progress__underlay" fill="transparent" cx="41.29032258064516" cy="41.29032258064516" r="20" stroke-width="1.2903225806451613" stroke-dasharray="125.664" stroke-dashoffset="0"></circle>
    <circle class="progress__overlay" fill="transparent" cx="41.29032258064516" cy="41.29032258064516" r="20" stroke-width="1.2903225806451613" stroke-dasharray="125.664" stroke-dashoffset="84.19468311620646px"></circle>
  </svg>
  <div class="progress__info"></div>
</div>
`;
